### PR TITLE
Add support for PHP 8.0

### PR DESCRIPTION
**Pull request by xvilo**
This does upgrade the code to allow for PHP 8.0, which can be merged as-is if #579 is something we do not want to add. However, if we want to add #579, then I can update this PR with the accompanying Rector changes :)

### CODE REVIEW

**Comment by xvilo** [comment]
@akondas can you tell me how buddy works, it seems it's not pickup the runner config change. According to the logs, the used CI image is still 7.4

**Comment by akondas** [comment]
I need to approve these changes, but there is also a problem with the public instance, it needs to be prepared for php 8

**Comment by akondas** [comment]
I have to think about how to do it so that it would not be downtime, I would have to install php 8 there beforehand, which is actually to be done

**Comment by akondas** [comment]
I still have something to finish this week, but from next week I will sit down to the repman and I will try to catch up with all PR (not only this one). We will catch up with this - I promise :wink:

**Comment by xvilo** [comment]
> I have to think about how to do it so that it would not be downtime

The best would then be to `"PHP": "7.4 || 8.0"`, merge this, update your vhost with the right PHP(-fpm) instance, and do a reload. Then we can fully move the codebase to 8.0, do you want me to update this one?

**Comment by akondas** [comment]
yes, good approach, that should allow the update to be done without downtime, thanks for help :+1:

**Comment by xvilo** [comment]
@akondas I've pushed the suggested update, just have to wait for the CI. If this is green, can you review and merge this PR?

**Comment by xvilo** [comment]
I'm unsure why the CI fails, it should be fine. It appears it might install the dependencies from some sort of cached vendor dir. I could use some help with this :)

File: composer.json
Comment (marmichalski) [inline composer.json:18]:
```suggestion
        "php": "^7.4 || ^8.0",
```

Comment (xvilo) [reply composer.json]: Good one, added this

File: composer.json
Comment (marmichalski) [inline composer.json:53]: AFAIK this update is not just as straightforward as presented (no changes to [the configuration](https://github.com/repman-io/repman/blob/a628a0a3969427dec8d444c5459bc2ec618e08c5/config/packages/prod/sentry.yaml)).

See `Removed the sentry.options.excluded_exceptions configuration option.`

* https://github.com/getsentry/sentry-symfony/blob/28ebfb37940e4f41e4fac9f0400e9e9439d1e4cc/UPGRADE-4.0.md

Comment (xvilo) [reply composer.json:53]: Ah good catch, I'll fix this tomorrow by using the intergration instead!

Comment (xvilo) [reply composer.json:53]: @marmichalski I've pushed support for this, can you do another review round?

File: composer.json
Comment (xvilo) [inline composer.json:28]: @akondas is there a specific reason to lock 1.3.0 before? We could use `^1.4` instead, which can be used to upgrade to any supported version in the 1x range.

Comment (akondas) [reply composer.json:28]: there is no specific reason, you can change this to `^1.4` :+1:

Comment (xvilo) [reply composer.json:28]: I've changed this, more easy updates in the future 👍

**Comment by xvilo** [comment]
@marmichalski / @akondas everything has been addressed, how is this PR looking now?

File: Dockerfile
Comment (xvilo) [reply Dockerfile:1]: I thought about doing this, but this step is more straight forward. And already caused some discussions. Let's do 8.0 first, then we can start working on 8.1 compatibility. This introduces more and stricter internal types. So possibly more stuffs to update 😀

**Comment by xvilo** [comment]
@akondas lets finish this?

File: Dockerfile
Comment (marmichalski) [inline Dockerfile:1]:
```suggestion
FROM php:8.0.19-fpm-alpine
```

File: buddy.yml
Comment (marmichalski) [inline buddy.yml:12]:
```suggestion
    docker_image_tag: "8.0.19"
```

File: buddy.yml
Comment (marmichalski) [inline buddy.yml:86]:
```suggestion
      docker_image_tag: "8.0.19"
```

File: composer.json
Comment (marmichalski) [inline composer.json:18]: So the plan is to drop 7.x in the next iteration? If not, how will the compatibility be guaranteed when CI is only run with one version? :thinking: @akondas?

**Review by marmichalski — COMMENTED** [review]
:+1:

**Comment by akondas** [comment]
So from my point of view, plan looks like this:
- allow to use php 8 in composer.json
- switch pipeline to php 8
- upgrade public instance to php 8
- upgrade code (composer.lock) to php 8
- drop suppoer for php 7.4

We are on first point here, so please do not run composer update on php 8.

**Comment by Jeroeny** [comment]
You can do this in composer.json to keep composer.lock depdencies on 7.4 for now:
```json
"config":
        "platform": {
            "php": "7.4"
        },
```

**Comment by xvilo** [comment]
I have applied @Jeroeny's suggestion and installed a 7.4 compatible set of dependencies.

**Review by marmichalski — APPROVED** [review]

**Comment by xvilo** [comment]
@akondas do you have some time to take a look at it?

**Comment by akondas** [comment]
Ok, let's go with this. Please resolve conflicts and we can merge this.

**Comment by xvilo** [comment]
@akondas merge conflicts are resolved, as soon as it is merged we will start using Repman on PHP 8.0 👍

**Comment by xvilo** [comment]
@akondas I've resolved the merge conflicts. Unfortunately the test pipeline is not passing due to some weird behaviour regarding PHPStan. I suspect it caches some files that has been corrupt, locally it does all work fine

**Comment by paulwehage** [comment]
Any news here? Would love to use repman with PHP 8.0.

**Comment by xvilo** [comment]
In the past 24 days there were new merge conflicts. From my side it's good to go, but @akondas needs to give final approval + merge. I've resolved the new conflicts, but I will not be actively watching the status anymore for now (it's taking long enough) and we're running these changes already in prod with our internal fork.

**Comment by xvilo** [comment]
@akondas I have, again, resolved the merge conflicts. Could we please merge this PR? We've been running this code on our on-premise install for the past 2-3 months and it runs fine :)

**Comment by 94noni** [comment]
@xvilo what about upgrading to [php8.1](https://www.php.net/supported-versions.php)?

**Comment by xvilo** [comment]
Good sugesstion, however, the upgrade path to 8.1 is a bit harder. I'd like to get this merged first. And honestly I'm wondering why this takes so long 😟

Also, an upgrade to Symfony 6.2 might be warranted. As this is also a fresh and hot release!

**Comment by xvilo** [comment]
@akondas We're closing in on a year, and have passed the EOL date of PHP 7.4. We've been running this code for almost a year with our private instance and haven't noticed any issues. 

Can we get this reviewed and merged?

**Comment by xvilo** [comment]
@akondas Whats your idea? Do you still want this or not?

**Comment by xvilo** [comment]
It seems the tests fails due to a small timing issue, I tried to quickly require symfony/clock, but I can't due to https://github.com/repman-io/repman/issues/653

**Comment by ymorocutti** [comment]
I was able to reproduce multiple times, but I'm not sure it's related to PHP 8.
My guess, you're having 2 DateTimes created within an interval of several milliseconds but are not on the same seconds.
ie : 10.980 vs 11.002.

@akondas Might be a good idea to merge this one asap, as PHP 7.4 has reach end of life and should no longer be used :)

**Comment by xvilo** [comment]
Yeah I expected so, but since repman was broken I kept it as is. Again, we’re using this as a fork on our kubernetes environment and all works fine as is! 

I would also like to work on PHP 8.1/2 compatibility and an upgrade to Symfony 6.3, but I will wait for this until we’re good with this one

**Comment by xvilo** [comment]
I thought we had this running internally already, but apparently we didn't yet. I've just switched our internal hosted version to this branch correctly and found one bug which is fixed with https://github.com/repman-io/repman/pull/580/commits/4ea3214aeac1c71299a6230bb271619ff445df61

For now I'll keep an eye on how well it works through Sentry

**Comment by xvilo** [comment]
@akondas in the past commits I've done the following:
- Add GitHub actions support for PHPUnit (just that for now, as Buddy is broken)
   Screenshot 2023-07-09 at 20 27 16
    - This currently runs tests both on PHP 7.4 and 8.0
- Update `coduo/php-matcher` as it indeed has incompatibilities when running PHPUnit tests
- Update a test to account for different wording between PHP 7.4 and PHP 8.0

As of know, both jobs report a succesfull status;
- 7.4: https://github.com/xvilo/repman/actions/runs/5501289892/jobs/10024774435
- 8.0: https://github.com/xvilo/repman/actions/runs/5501289892/jobs/10024774499

File: .github/workflows/unit-tests.yml
Comment (marmichalski) [inline .github/workflows/unit-tests.yml:11]:
```suggestion
                image: postgres:11.7-alpine
```

I think you want to test against what's in docker-compose - it's a little bit old though. Maybe there should be a matrix for postgres versions, too? 🙈

File: .github/workflows/unit-tests.yml
Comment (marmichalski) [inline .github/workflows/unit-tests.yml:18]:
```suggestion
                options: --health-cmd pg_isready --health-interval 1s --health-timeout 3s --health-retries 3
```

no need to wait that long

File: .github/workflows/unit-tests.yml
Comment (marmichalski) [inline .github/workflows/unit-tests.yml:5]:
```suggestion
        name: PHPUnit tests (PHP ${{ matrix.php-version }})
```

File: .github/workflows/unit-tests.yml
Comment (marmichalski) [inline .github/workflows/unit-tests.yml:22]:
```suggestion
                php-version: ['7.4', '8.0']
```

File: .github/workflows/unit-tests.yml
Comment (marmichalski) [inline .github/workflows/unit-tests.yml:31]:
```suggestion
                  php-version: ${{ matrix.php-version }}
```

File: .github/workflows/unit-tests.yml
Comment (marmichalski) [inline .github/workflows/unit-tests.yml:40]:
```suggestion

```

File: .github/workflows/unit-tests.yml
Comment (marmichalski) [inline .github/workflows/unit-tests.yml:53]:
```suggestion

```

don't see how that's relevant?

File: .github/workflows/unit-tests.yml
Comment (marmichalski) [inline .github/workflows/unit-tests.yml:59]:
```suggestion
```

File: .github/workflows/unit-tests.yml
Comment (marmichalski) [inline .github/workflows/unit-tests.yml:10]: Why is this needed for unit tests? Did you mean to name the file phpunit-tests?

File: .github/workflows/unit-tests.yml
Comment (marmichalski) [inline .github/workflows/unit-tests.yml:70]:
```suggestion
                  php bin/console doctrine:migrations:migrate --no-interaction"
                  php bin/console messenger:setup-transports"
```

I think you want those to fail properly... 🤔

File: buddy.yml
Comment (marmichalski) [inline buddy.yml:15]:
```suggestion
    - "composer validate"
    - "composer install"
```

File: config/services.yaml
Comment (marmichalski) [inline config/services.yaml:132]: I assume all of the optional-to-non-optional changes are because the services are now always registered?

**Review by marmichalski — APPROVED** [review]
Guess that's alright. Workflow could get some love

File: config/services.yaml
Comment (xvilo) [reply config/services.yaml:132]: Yes they are indeed, I found at that the wrong service was getting injected, so I figured it would be best to just specifically point to the interfaces instead

File: .github/workflows/unit-tests.yml
Comment (xvilo) [reply .github/workflows/unit-tests.yml:10]: The PHPUnit tests also run functional tests, so it's not just "unit-tests", I will rename the file and job

File: .github/workflows/unit-tests.yml
Comment (xvilo) [reply .github/workflows/unit-tests.yml:53]: Well there is a comment on top, describing the issue. Within phpunit and the functional tests we actually create a new git repo and without this we can not run `git add .` (https://github.com/repman-io/repman/blob/c384362816510d267488773077f4dbb73594c035/tests/Unit/Service/Security/SensioLabsSecurityCheckerTest.php#L156C1-L156C1)  it will result in exit code 128 and the next steps just fail silently

Comment (marmichalski) [reply .github/workflows/unit-tests.yml]: Oooh, I see, sorry! Thought it was something around workflow, not the actual tests using `git add` :+1:

File: .github/workflows/unit-tests.yml
Comment (xvilo) [reply .github/workflows/unit-tests.yml]: Yeah good one, added `'11.20-alpine', '14.8-alpine', '15.3-alpine'` some random versions

**Comment by xvilo** [comment]
So I made small booboo and accidentlly pushed custom patches this way, cleaned and cherry-picked it. 

Also thanks to @marmichalski for the review, answered and/or fixed the comments you left

PHP Unit tests statusses all in the green:
- Postgres 11.20:
    - PHP 7.4: https://github.com/xvilo/repman/actions/runs/5506320670/jobs/10034839156
    - PHP 8.0: https://github.com/xvilo/repman/actions/runs/5506320670/jobs/10034839692
- Postgres 13.11:
    - PHP 7.4: https://github.com/xvilo/repman/actions/runs/5506320670/jobs/10034839366
    - PHP 8.0: https://github.com/xvilo/repman/actions/runs/5506320670/jobs/10034839838
- Postgres 15.03:
    - PHP 7.4: https://github.com/xvilo/repman/actions/runs/5506320670/jobs/10034839517
    - PHP 8.0: https://github.com/xvilo/repman/actions/runs/5506320670/jobs/10034840032

**Comment by xvilo** [comment]
@akondas ?

**Comment by xvilo** [comment]
@akondas do you have some time to look into this MR? PHP 8.0 is EOL this week, I would like to get this merged and add some effort in upgrading it to newer php versions. Also resolving some bugs we're hitting, but I will not invest more time until I get some response on here 👍

**Review by marmichalski — APPROVED** [review]

**Review by Jeroeny — APPROVED** [review]

**Comment by xvilo** [comment]
Thanks for picking this up @marmichalski, let me know if there is anything I can do to help out 👍

**Comment by marmichalski** [comment]
> Thanks for picking this up @marmichalski, let me know if there is anything I can do to help out 👍

Yea, I have no idea why some tests under `ComposerPackageSynchronizer` are failing in PHP 8.0 CI, considering they pass in 7.4 CI and locally for me 🤷

**Comment by xvilo** [comment]
That’s weird, in one of my previous comments I had both 7.4 and 8.0 passing through GitHub actions just fine. I’ll take a look tomorrow or later this week if you haven’t solved it yet

**Comment by marmichalski** [comment]
> That’s weird, in one of my previous comments I had both 7.4 and 8.0 passing through GitHub actions just fine. I’ll take a look tomorrow or later this week if you haven’t solved it yet.

I've brought your original workflow to life and even 7.4 failed on those tests 🙈

**Comment by xvilo** [comment]
Locall on macos I have the following 2 failing:
```
PHPUnit 9.5.25 #StandWithUkraine

...............................................................  63 / 500 ( 12%)
............................................................... 126 / 500 ( 25%)
........................E.E.................................... 189 / 500 ( 37%)
............................................................... 252 / 500 ( 50%)
............................................................... 315 / 500 ( 63%)
............................................................... 378 / 500 ( 75%)
............................................................... 441 / 500 ( 88%)
...........................................................     500 / 500 (100%)

You should really speed up these slow tests (>500ms)...
 1. 514ms to run Buddy\\Repman\\Tests\\Functional\\Controller\\Admin\\ConfigControllerTest::testToggleAuthenticationOptions


Time: 00:10.786, Memory: 141.00 MB

There were 2 errors:

1) Buddy\Repman\Tests\Unit\Service\Security\SensioLabsSecurityCheckerTest::testUpdateWhenRepoDontExist
Symfony\Component\Process\Exception\ProcessFailedException: The command "'git' 'clone' '--depth' '1' '--branch' 'master' 'file:///var/folders/zj/l6h4qqtn7hx6hcz2v4l9wl4m0000gn/T/repman/security-advisories-repo' '.'" failed.

Exit Code: 128(Invalid exit argument)

Working directory: /var/folders/zj/l6h4qqtn7hx6hcz2v4l9wl4m0000gn/T/repman/security-advisories

Output:
================


Error Output:
================
Cloning into '.'...
warning: Could not find remote branch master to clone.
fatal: Remote branch master not found in upstream origin


/Users/xvilo/projects/repman/src/Service/Security/SecurityChecker/SensioLabsSecurityChecker.php:218
/Users/xvilo/projects/repman/src/Service/Security/SecurityChecker/SensioLabsSecurityChecker.php:196
/Users/xvilo/projects/repman/src/Service/Security/SecurityChecker/SensioLabsSecurityChecker.php:34
/Users/xvilo/projects/repman/tests/Unit/Service/Security/SensioLabsSecurityCheckerTest.php:119

2) Buddy\Repman\Tests\Unit\Service\Security\SensioLabsSecurityCheckerTest::testUpdateWhenRepoExist
Symfony\Component\Process\Exception\ProcessFailedException: The command "'git' 'clone' '--depth' '1' '--branch' 'master' 'file:///var/folders/zj/l6h4qqtn7hx6hcz2v4l9wl4m0000gn/T/repman/security-advisories-repo' '.'" failed.

Exit Code: 128(Invalid exit argument)

Working directory: /var/folders/zj/l6h4qqtn7hx6hcz2v4l9wl4m0000gn/T/repman/security-advisories

Output:
================


Error Output:
================
Cloning into '.'...
warning: Could not find remote branch master to clone.
fatal: Remote branch master not found in upstream origin


/Users/xvilo/projects/repman/src/Service/Security/SecurityChecker/SensioLabsSecurityChecker.php:218
/Users/xvilo/projects/repman/src/Service/Security/SecurityChecker/SensioLabsSecurityChecker.php:196
/Users/xvilo/projects/repman/src/Service/Security/SecurityChecker/SensioLabsSecurityChecker.php:34
/Users/xvilo/projects/repman/tests/Unit/Service/Security/SensioLabsSecurityCheckerTest.php:134

ERRORS!
Tests: 500, Assertions: 1201, Errors: 2.
```

**Comment by marmichalski** [comment]
Yea, those are the reason of macOS git config (not the git version, as newer git versions are still using `master` as [default branch](https://git-scm.com/docs/git-init#Documentation/git-init.txt---initial-branchltbranch-namegt)). Assuming that would work for you if you'd init the repo with `--initial-branch master` in that test, so everything would pass on your local, too :sob: 

I am still considering how to properly go around this; I think we don't have to specify branch at all in `SensioLabsSecurityChecker` as `git clone` with `--depth 1` will fetch single (default) branch anyway (IIRC).

**Comment by xvilo** [comment]
From what I have found so far it that it appears to read the versions of the project itself in the test? 
The buffered IO contains the following logs:
```
Reading composer.json of repman-io/repman (0.1.0)\n
Importing tag 0.1.0 (0.1.0.0)\n
Reading composer.json of repman-io/repman (0.1.1)\n
Importing tag 0.1.1 (0.1.1.0)\n
Reading composer.json of repman-io/repman (0.1.2)\n
Importing tag 0.1.2 (0.1.2.0)\n
Reading composer.json of repman-io/repman (0.2.0)\n
Importing tag 0.2.0 (0.2.0.0)\n
Reading composer.json of repman-io/repman (0.2.1)\n
Importing tag 0.2.1 (0.2.1.0)\n
Reading composer.json of repman-io/repman (0.3.0)\n
Importing tag 0.3.0 (0.3.0.0)\n
Reading composer.json of repman-io/repman (0.4.0)\n
Importing tag 0.4.0 (0.4.0.0)\n
Reading composer.json of repman-io/repman (0.4.1)\n
Importing tag 0.4.1 (0.4.1.0)\n
Reading composer.json of repman-io/repman (0.5.0)\n
Importing tag 0.5.0 (0.5.0.0)\n
Reading composer.json of repman-io/repman (0.6.0)\n
Importing tag 0.6.0 (0.6.0.0)\n
Reading composer.json of repman-io/repman (1.0.0)\n
Importing tag 1.0.0 (1.0.0.0)\n
Reading composer.json of repman-io/repman (1.1.0)\n
Importing tag 1.1.0 (1.1.0.0)\n
Reading composer.json of repman-io/repman (1.1.1)\n
Importing tag 1.1.1 (1.1.1.0)\n
Reading composer.json of repman-io/repman (1.2.0)\n
Importing tag 1.2.0 (1.2.0.0)\n
Reading composer.json of repman-io/repman (1.2.1)\n
Importing tag 1.2.1 (1.2.1.0)\n
Reading composer.json of repman-io/repman (1.2.2)\n
Importing tag 1.2.2 (1.2.2.0)\n
Reading composer.json of repman-io/repman (1.3.0)\n
Importing tag 1.3.0 (1.3.0.0)\n
Reading composer.json of repman-io/repman (1.3.1)\n
Importing tag 1.3.1 (1.3.1.0)\n
Reading composer.json of repman-io/repman (1.3.2)\n
Importing tag 1.3.2 (1.3.2.0)\n
Reading composer.json of repman-io/repman (1.3.3)\n
Importing tag 1.3.3 (1.3.3.0)\n
Reading composer.json of repman-io/repman (1.3.4)\n
Importing tag 1.3.4 (1.3.4.0)\n
Reading composer.json of repman-io/repman (1.3.5)\n
Importing tag 1.3.5 (1.3.5.0)\n
Reading composer.json of repman-io/repman (1.3.6)\n
Importing tag 1.3.6 (1.3.6.0)\n
Reading composer.json of repman-io/repman (feat/php-8.0)\n
Importing branch feat/php-8.0 (dev-feat/php-8.0)\n
Reading composer.json of repman-io/repman (master)\n
Importing branch master (dev-master)\n
```

Those are clearly the branches and tags of this project, and not, as I expected, the versions of the dists zips in the tests folder. I'll check if the checkout is shallow or something like that

**Comment by xvilo** [comment]
That was it! I've added the `fetch-depth` and `fetch-tags` to the `actions/checkout@v3` action. This makes it so the copy isn't (too) shallow. I think it shouldn't fetch itself, so the test might be incorrect. However, I believe this was already the case so I wont address that in this PR. 

@marmichalski how you like?

**Comment by marmichalski** [comment]
> That was it! I've added the `fetch-depth` and `fetch-tags` to the `actions/checkout@v3` action. This makes it so the copy isn't (too) shallow. I think it shouldn't fetch itself, so the test might be incorrect. However, I believe this was already the case so I wont address that in this PR.
> 
> @marmichalski how you like?

Can you make the other workflow pass and drop this one, though? :+1: 
I don't think we should be fetching full history at all times, maybe `fetch-tags: true` would be enough? Also this does not really explain why the workflow (the one I'd like to keep) is passing on 7.4 and not 8.0

**Comment by xvilo** [comment]
@marmichalski ah, I've been debugging the wrong file in that case, as that is a different error. I'll restart my investigation with that one

**Comment by xvilo** [comment]
It might be an issue with git safe directories. I would assume that the 7.4 image might be a bit more outdated then the 8.0 image

**Comment by marmichalski** [comment]
> It might be an issue with git safe directories. I would assume that the 7.4 image might be a bit more outdated then the 8.0 image

Yea, looks to be the case looking at the dubious ownership. Good find :detective:

File: .github/workflows/tests.yaml
Comment (marmichalski) [inline .github/workflows/tests.yaml:67]:
```suggestion
```

**Review by marmichalski — APPROVED** [review]

**Comment by xvilo** [comment]
I think it can be merged in to the main branch as is, I don't see any objections right now. We've been using this in prod for a while now

**Comment by marmichalski** [comment]
> I think it can be merged in to the main branch as is, I don't see any objections right now. We've been using this in prod for a while now through a custom fork

Thank you, let's see what happens next :crossed_fingers:

**Comment by xvilo** [comment]
> > I think it can be merged in to the main branch as is, I don't see any objections right now. We've been using this in prod for a while now through a custom fork
> 
> Thank you, let's see what happens next 🤞

Do you want me to find some time to work on php 8.2 compatibility, or will you handle this? PR was already open (https://github.com/repman-io/repman/pull/660), but with some failing tests that I didn't know enough about

**Comment by marmichalski** [comment]
> Do you want me to find some time to work on php 8.2 compatibility, or will you handle this? PR was already open (#660), but with some failing tests that I didn't know enough about

I can take over that PR in my spare time :)

**Pull request merged**
